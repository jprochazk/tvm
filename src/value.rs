use std::fmt::{Debug, Display};
use std::hash::Hash;
use std::str::FromStr;

use crate::{HashMap, Str};

#[derive(Default)]
pub struct ConstantPoolBuilder<'src> {
    ints: HashMap<i64, u16>,
    nums: HashMap<f64n, u16>,
    strings: HashMap<Str<'src>, u16>,
    jump_offsets: HashMap<isize, u16>,
    pool: Vec<Constant>,
}

impl<'src> ConstantPoolBuilder<'src> {
    pub fn new() -> Self {
        Self {
            ints: HashMap::default(),
            nums: HashMap::default(),
            strings: HashMap::default(),
            jump_offsets: HashMap::default(),
            pool: Vec::new(),
        }
    }

    pub fn insert_int(&mut self, v: i64) -> Option<u16> {
        insert_or_get_in(&mut self.ints, &mut self.pool, v)
    }

    pub fn insert_num(&mut self, v: f64n) -> Option<u16> {
        insert_or_get_in(&mut self.nums, &mut self.pool, v)
    }

    pub fn insert_str(&mut self, v: Str<'src>) -> Option<u16> {
        insert_or_get_in(&mut self.strings, &mut self.pool, v)
    }

    pub fn insert_jump_offset(&mut self, v: isize) -> Option<u16> {
        insert_or_get_in(&mut self.jump_offsets, &mut self.pool, v)
    }

    pub fn finish(self) -> Vec<Constant> {
        self.pool
    }
}

fn insert_or_get_in<T: Clone + Eq + Hash + ToConstant>(
    m: &mut HashMap<T, u16>,
    p: &mut Vec<Constant>,
    v: T,
) -> Option<u16> {
    // too many constants
    if p.len() >= u16::MAX as usize {
        return None;
    }

    use std::collections::hash_map::Entry as E;
    let idx = match m.entry(v.clone()) {
        E::Occupied(e) => *e.get(),
        E::Vacant(e) => {
            let idx = *e.insert(p.len() as u16);
            p.push(v.to_constant());
            idx
        }
    };
    Some(idx)
}

#[derive(Debug)]
pub enum Constant {
    Int(i64),
    Num(f64n),
    Str(String),
    Jmp(isize),
}

/// An `f64` which is guaranteed to be non-NaN
#[allow(non_camel_case_types)]
#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
#[repr(transparent)]
pub struct f64n {
    value: u64,
}

impl f64n {
    /// # Panics
    /// If `value.is_nan()`
    pub fn new(value: f64) -> Self {
        assert!(!value.is_nan());

        Self {
            value: value.to_bits(),
        }
    }

    #[inline]
    pub fn get(self) -> f64 {
        f64::from_bits(self.value)
    }
}

impl Debug for f64n {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        Debug::fmt(&self.get(), f)
    }
}

impl Display for f64n {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        Display::fmt(&self.get(), f)
    }
}

pub struct ParseFloatError {
    inner: ParseFloatErrorInner,
}

enum ParseFloatErrorInner {
    Std(std::num::ParseFloatError),
    Nan,
}

impl Debug for ParseFloatError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match &self.inner {
            ParseFloatErrorInner::Std(e) => Debug::fmt(e, f),
            ParseFloatErrorInner::Nan => f
                .debug_struct("ParseFloatError")
                .field("kind", &"NaN")
                .finish(),
        }
    }
}

impl Display for ParseFloatError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match &self.inner {
            ParseFloatErrorInner::Std(e) => Display::fmt(e, f),
            ParseFloatErrorInner::Nan => f.write_str("float literal must not be nan"),
        }
    }
}

impl std::error::Error for ParseFloatError {}

impl FromStr for f64n {
    type Err = ParseFloatError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let v = f64::from_str(s).map_err(|e| ParseFloatError {
            inner: ParseFloatErrorInner::Std(e),
        })?;
        if v.is_nan() {
            return Err(ParseFloatError {
                inner: ParseFloatErrorInner::Nan,
            });
        }
        Ok(f64n { value: v.to_bits() })
    }
}

trait ToConstant {
    fn to_constant(self) -> Constant;
}

impl ToConstant for i64 {
    fn to_constant(self) -> Constant {
        Constant::Int(self)
    }
}

impl ToConstant for f64n {
    fn to_constant(self) -> Constant {
        Constant::Num(self)
    }
}

impl ToConstant for Str<'_> {
    fn to_constant(self) -> Constant {
        Constant::Str(self.to_string())
    }
}

impl ToConstant for isize {
    fn to_constant(self) -> Constant {
        Constant::Jmp(self)
    }
}
