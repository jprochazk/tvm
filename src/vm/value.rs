use super::{ExternFunctionError, TryFromValue, ValueAbi};
use crate::hir;

pub mod intern;
pub mod pool;

#[derive(Default, Debug, Clone, Copy)]
#[repr(u64)]
pub enum Value {
    #[default]
    Unit = 0,
    I64(i64) = 1,
    F64(f64) = 2,
    Bool(bool) = 3,
}

impl Value {
    #[inline]
    pub fn tag(self) -> u64 {
        // SAFETY: Because `Self` is marked `repr(u64)`, its layout is a `repr(C)`
        // `union` between `repr(C)` structs, each of which has the `u64`
        // discriminant as its first field, so we can read the discriminant
        // without offsetting the pointer.
        unsafe {
            let ptr = &self as *const Self as *const u64;
            ptr.read()
        }
    }

    #[inline]
    pub fn unit(self) -> Option<()> {
        match self {
            Self::Unit => Some(()),
            _ => None,
        }
    }

    /// # Safety
    /// - `self` must be `Self::Unit`
    #[inline]
    pub unsafe fn unit_unchecked(self) {
        match self {
            Value::Unit => (),
            _ => core::hint::unreachable_unchecked(),
        }
    }

    #[inline]
    pub fn i64(self) -> Option<i64> {
        match self {
            Self::I64(v) => Some(v),
            _ => None,
        }
    }

    /// # Safety
    /// - `self` must be `Self::I64`
    #[inline]
    pub unsafe fn i64_unchecked(self) -> i64 {
        assert!(matches!(self, Self::I64(_)));
        match self {
            Value::I64(v) => v,
            _ => core::hint::unreachable_unchecked(),
        }
    }

    #[inline]
    pub fn f64(self) -> Option<f64> {
        match self {
            Self::F64(v) => Some(v),
            _ => None,
        }
    }

    /// # Safety
    /// - `self` must be `Self::F64`
    #[inline]
    pub unsafe fn f64_unchecked(self) -> f64 {
        debug_assert!(matches!(self, Self::F64(_)));
        match self {
            Value::F64(v) => v,
            _ => core::hint::unreachable_unchecked(),
        }
    }

    #[inline]
    pub fn bool(self) -> Option<bool> {
        match self {
            Self::Bool(v) => Some(v),
            _ => None,
        }
    }

    /// # Safety
    /// - `self` must be `Self::Bool`
    #[inline]
    pub unsafe fn bool_unchecked(self) -> bool {
        debug_assert!(matches!(self, Self::Bool(_)));
        match self {
            Value::Bool(v) => v,
            _ => core::hint::unreachable_unchecked(),
        }
    }
}

impl TryFromValue for Value {
    #[inline]
    fn try_from_value(value: Value) -> std::result::Result<Self, ExternFunctionError> {
        Ok(value)
    }
}

impl ValueAbi for Value {
    const TYPE: hir::Ty = hir::Ty::Dynamic;
}

impl From<()> for Value {
    #[inline]
    fn from(_: ()) -> Self {
        Self::Unit
    }
}

impl TryFromValue for () {
    #[inline]
    fn try_from_value(value: Value) -> std::result::Result<Self, ExternFunctionError> {
        match value {
            Value::Unit => Ok(()),
            // TODO: proper error
            _ => Err(()),
        }
    }
}

impl ValueAbi for () {
    const TYPE: hir::Ty = hir::Ty::Unit;
}

impl From<i64> for Value {
    #[inline]
    fn from(value: i64) -> Self {
        Self::I64(value)
    }
}

impl TryFromValue for i64 {
    #[inline]
    fn try_from_value(value: Value) -> std::result::Result<Self, ExternFunctionError> {
        match value {
            Value::I64(v) => Ok(v),
            // TODO: proper error
            _ => Err(()),
        }
    }
}

impl ValueAbi for i64 {
    const TYPE: hir::Ty = hir::Ty::INT;
}

impl From<f64> for Value {
    #[inline]
    fn from(value: f64) -> Self {
        Self::F64(value)
    }
}

impl TryFromValue for f64 {
    #[inline]
    fn try_from_value(value: Value) -> std::result::Result<Self, ExternFunctionError> {
        match value {
            Value::F64(v) => Ok(v),
            // TODO: proper error
            _ => Err(()),
        }
    }
}

impl ValueAbi for f64 {
    const TYPE: hir::Ty = hir::Ty::NUM;
}

impl From<bool> for Value {
    #[inline]
    fn from(value: bool) -> Self {
        Self::Bool(value)
    }
}

impl TryFromValue for bool {
    #[inline]
    fn try_from_value(value: Value) -> std::result::Result<Self, ExternFunctionError> {
        match value {
            Value::Bool(v) => Ok(v),
            // TODO: proper error
            _ => Err(()),
        }
    }
}

impl ValueAbi for bool {
    const TYPE: hir::Ty = hir::Ty::BOOL;
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
#[repr(u64)]
pub enum Literal {
    JumpOffset(isize) = 0,
    I64(i64) = 1,
    F64(f64n) = 2,
}

impl Literal {
    /// # Safety
    /// - `literal` must not be `Literal::Jmp`
    #[inline]
    pub unsafe fn into_value(self) -> Value {
        debug_assert!(!matches!(self, Literal::JumpOffset(_)));
        match self {
            Literal::JumpOffset(_) => unsafe { core::hint::unreachable_unchecked() },
            Literal::I64(v) => Value::I64(v),
            Literal::F64(v) => Value::F64(v.into()),
        }
    }

    #[inline]
    pub fn tag(self) -> u64 {
        // SAFETY: Because `Self` is marked `repr(u64)`, its layout is a `repr(C)`
        // `union` between `repr(C)` structs, each of which has the `u64`
        // discriminant as its first field, so we can read the discriminant
        // without offsetting the pointer.
        unsafe {
            let ptr = &self as *const Self as *const u64;
            ptr.read()
        }
    }

    #[inline]
    pub fn i64(self) -> Option<i64> {
        match self {
            Self::I64(v) => Some(v),
            _ => None,
        }
    }

    /// # Safety
    /// - `self` must be `Literal::I64`
    #[inline]
    pub unsafe fn i64_unchecked(self) -> i64 {
        debug_assert!(!matches!(self, Literal::JumpOffset(_)));
        match self {
            Literal::I64(v) => v,
            _ => core::hint::unreachable_unchecked(),
        }
    }

    #[inline]
    pub fn f64(self) -> Option<f64> {
        match self {
            Self::F64(v) => Some(v.get()),
            _ => None,
        }
    }

    /// # Safety
    /// - `self` must be `Literal::F64`
    #[inline]
    pub unsafe fn f64_unchecked(self) -> f64 {
        debug_assert!(!matches!(self, Literal::JumpOffset(_)));
        match self {
            Literal::F64(v) => v.get(),
            _ => core::hint::unreachable_unchecked(),
        }
    }

    #[inline]
    pub fn jump_offset(self) -> Option<isize> {
        match self {
            Self::JumpOffset(v) => Some(v),
            _ => None,
        }
    }

    /// # Safety
    /// - `self` must be `Literal::JumpOffset`
    #[inline]
    pub unsafe fn jump_offset_unchecked(self) -> isize {
        debug_assert!(matches!(self, Literal::JumpOffset(_)));
        match self {
            Literal::JumpOffset(v) => v,
            _ => core::hint::unreachable_unchecked(),
        }
    }
}

impl Literal {
    #[inline]
    pub fn jmp(offset: isize) -> Self {
        Self::JumpOffset(offset)
    }
}

impl From<i64> for Literal {
    #[inline]
    fn from(value: i64) -> Self {
        Self::I64(value)
    }
}

impl From<f64n> for Literal {
    #[inline]
    fn from(value: f64n) -> Self {
        Self::F64(value)
    }
}

/// An `f64` which is guaranteed to be non-NaN
#[allow(non_camel_case_types)]
#[derive(Clone, Copy)]
#[repr(transparent)]
pub struct f64n {
    value: f64,
}

impl f64n {
    pub const ZERO: f64n = f64n { value: 0.0 };

    /// # Panics
    /// If `value.is_nan()`
    #[inline]
    pub fn new(value: f64) -> Self {
        assert!(!value.is_nan());

        Self { value }
    }

    #[inline]
    pub fn try_new(value: f64) -> Option<Self> {
        if value.is_nan() {
            return None;
        }

        Some(Self { value })
    }

    #[inline]
    pub fn get(self) -> f64 {
        self.value
    }
}

impl std::fmt::Debug for f64n {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        std::fmt::Debug::fmt(&self.get(), f)
    }
}

impl std::fmt::Display for f64n {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        std::fmt::Display::fmt(&self.get(), f)
    }
}

impl PartialEq for f64n {
    #[inline]
    fn eq(&self, other: &Self) -> bool {
        self.value.to_bits() == other.value.to_bits()
    }
}

impl Eq for f64n {}

impl PartialOrd for f64n {
    #[inline]
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for f64n {
    #[inline]
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.value.to_bits().cmp(&other.value.to_bits())
    }
}

impl std::hash::Hash for f64n {
    #[inline]
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.value.to_bits().hash(state);
    }
}

impl From<f64n> for f64 {
    #[inline]
    fn from(value: f64n) -> Self {
        value.get()
    }
}

pub struct ParseFloatError {
    inner: ParseFloatErrorInner,
}

enum ParseFloatErrorInner {
    Std(std::num::ParseFloatError),
    Nan,
}

impl std::fmt::Debug for ParseFloatError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match &self.inner {
            ParseFloatErrorInner::Std(e) => std::fmt::Debug::fmt(e, f),
            ParseFloatErrorInner::Nan => f
                .debug_struct("ParseFloatError")
                .field("kind", &"NaN")
                .finish(),
        }
    }
}

impl std::fmt::Display for ParseFloatError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match &self.inner {
            ParseFloatErrorInner::Std(e) => std::fmt::Display::fmt(e, f),
            ParseFloatErrorInner::Nan => f.write_str("float literal must not be nan"),
        }
    }
}

impl std::error::Error for ParseFloatError {}

impl From<std::num::ParseFloatError> for ParseFloatError {
    #[inline]
    fn from(value: std::num::ParseFloatError) -> Self {
        ParseFloatError {
            inner: ParseFloatErrorInner::Std(value),
        }
    }
}

impl std::str::FromStr for f64n {
    type Err = ParseFloatError;

    #[inline]
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let v = f64::from_str(s)?;
        f64n::try_new(v).ok_or(ParseFloatError {
            inner: ParseFloatErrorInner::Nan,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn literal_value_tags_match() {
        assert_eq!(Literal::I64(0).tag(), Value::I64(0).tag());
        assert_eq!(Literal::F64(f64n::ZERO).tag(), Value::F64(0.0).tag());
    }
}
