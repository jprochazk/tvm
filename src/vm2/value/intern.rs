use std::hash::{Hash, Hasher};
use std::num::NonZeroU32;

use hashbrown::hash_map::RawEntryMut;

#[repr(transparent)]
#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Symbol(NonZeroU32);

impl Symbol {
    #[inline]
    fn from_index(index: usize) -> Self {
        Symbol(unsafe { NonZeroU32::new_unchecked(index as u32 + 1) })
    }

    #[inline]
    fn index(&self) -> usize {
        self.0.get() as usize - 1
    }
}

/// Used to ensure that none of the `HashMap` methods
/// which perform hashing internally are used, as we
/// need to do all hashing/equality operations manually.
#[repr(transparent)]
#[derive(Clone, Copy)]
struct NoHash<T>(T);

pub struct Interner {
    table: hashbrown::HashMap<NoHash<Symbol>, (), ()>,
    /// Indexed by `Symbol`, stores strings prefixed by length.
    buffer: Vec<u8>,
}

#[inline]
fn fxhash_str_bytes(s: &[u8]) -> u64 {
    let mut hasher = rustc_hash::FxHasher::default();
    hasher.write(s);
    hasher.write_u8(0xFF); // maintain prefix-freedom
    hasher.finish()
}

/// # Safety
/// - `symbol` must have been created by a call to `intern` that holds the given
///   `buffer`.
#[inline]
unsafe fn get_bytes_unchecked(buffer: &[u8], symbol: Symbol) -> &[u8] {
    let idx = symbol.index();
    // SAFETY: The indexing here is guaranteed to not be
    //         out of bounds for the buffer.
    debug_assert!(idx + 4 < buffer.len(), "symbol out of bounds");
    let len = buffer.get_unchecked(idx..idx + 4);
    let len = u32::from_le_bytes(len.try_into().unwrap()) as usize;
    debug_assert!(idx + 4 + len <= buffer.len(), "symbol out of bounds");
    buffer.get_unchecked(idx + 4..idx + 4 + len)
}

#[inline]
fn get_bytes(buffer: &[u8], symbol: Symbol) -> Option<&[u8]> {
    let idx = symbol.index();
    let len = buffer.get(idx..idx + 4)?;
    let len = u32::from_le_bytes(len.try_into().unwrap()) as usize;
    buffer.get(idx + 4..idx + 4 + len)
}

impl Interner {
    #[inline]
    pub fn new() -> Self {
        Self {
            table: hashbrown::HashMap::with_hasher(()),
            buffer: Vec::new(),
        }
    }

    pub fn intern(&mut self, data: &str) -> Symbol {
        let hash = fxhash_str_bytes(data.as_bytes());
        let entry = self
            .table
            .raw_entry_mut()
            .from_hash(hash, |&NoHash(symbol)| unsafe {
                get_bytes_unchecked(&self.buffer, symbol) == data.as_bytes()
            });
        match entry {
            RawEntryMut::Occupied(entry) => entry.key().0,
            RawEntryMut::Vacant(entry) => {
                let symbol = Symbol::from_index(self.buffer.len());
                self.buffer
                    .extend_from_slice(&(data.len() as u32).to_le_bytes());
                self.buffer.extend_from_slice(data.as_bytes());

                let (key, _) =
                    entry.insert_with_hasher(hash, NoHash(symbol), (), |&NoHash(symbol)| unsafe {
                        fxhash_str_bytes(get_bytes_unchecked(&self.buffer, symbol))
                    });

                key.0
            }
        }
    }

    #[inline]
    pub fn get(&self, symbol: Symbol) -> &str {
        let Some(bytes) = get_bytes(&self.buffer, symbol) else {
            panic!("symbol used with different interner");
        };

        let Ok(str) = std::str::from_utf8(bytes) else {
            panic!("symbol used with different interner");
        };

        str
    }

    /// # Safety
    /// - `symbol` must have been created by a call to `intern` that holds the
    ///   given `buffer`.
    #[inline]
    pub unsafe fn get_unchecked(&self, symbol: Symbol) -> &str {
        std::str::from_utf8_unchecked(get_bytes_unchecked(&self.buffer, symbol))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn intern_stuff() {
        let mut interner = Interner::new();
        let id = interner.intern("test");
        assert_eq!(interner.get(id), "test");
    }
}
