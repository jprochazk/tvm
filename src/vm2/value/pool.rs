use super::Literal;
use crate::vm2::operands::LiteralId;

#[derive(Default)]
pub struct LiteralPool {
    inner: Vec<Literal>,
    table: hashbrown::HashMap<LiteralId, (), ()>,
}

fn hash_literal(literal: &Literal) -> u64 {
    use std::hash::{Hash, Hasher};

    let mut hasher = rustc_hash::FxHasher::default();
    literal.hash(&mut hasher);
    hasher.finish()
}

impl LiteralPool {
    pub fn new() -> Self {
        Self {
            inner: Vec::new(),
            table: hashbrown::HashMap::default(),
        }
    }

    /// Returns `None` if the literal pool is full.
    pub fn insert(&mut self, literal: impl Into<Literal>) -> Option<LiteralId> {
        let literal = literal.into();

        let hash = hash_literal(&literal);
        let entry = self
            .table
            .raw_entry_mut()
            .from_hash(hash, |&id| literal == self.inner[id as usize]);
        match entry {
            hashbrown::hash_map::RawEntryMut::Occupied(entry) => Some(*entry.key()),
            hashbrown::hash_map::RawEntryMut::Vacant(raw_vacant_entry_mut) => {
                if self.inner.len() + 1 > u16::MAX as usize {
                    return None;
                }
                let id = self.inner.len() as u16;
                raw_vacant_entry_mut
                    .insert_with_hasher(hash, id, (), |&id| hash_literal(&self.inner[id as usize]));
                self.inner.push(literal);
                Some(id)
            }
        }
    }

    pub fn finish(self) -> Vec<Literal> {
        self.inner
    }
}
