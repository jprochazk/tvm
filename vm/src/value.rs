#[derive(Default, Debug, Clone, Copy)]
#[repr(u64)]
pub enum Value {
    #[default]
    Unit = 0,
    I64(i64) = 1,
    F64(f64) = 2,
    Bool(bool) = 3,
}

impl Value {
    #[inline]
    pub fn tag(self) -> u64 {
        // SAFETY: Because `Self` is marked `repr(u64)`, its layout is a `repr(C)`
        // `union` between `repr(C)` structs, each of which has the `u64`
        // discriminant as its first field, so we can read the discriminant
        // without offsetting the pointer.
        unsafe {
            let ptr = &self as *const Self as *const u64;
            ptr.read()
        }
    }

    #[inline]
    pub fn unit(self) -> Option<()> {
        match self {
            Self::Unit => Some(()),
            _ => None,
        }
    }

    /// # Safety
    /// - `self` must be `Self::Unit`
    #[inline]
    pub unsafe fn unit_unchecked(self) {
        match self {
            Value::Unit => (),
            _ => core::hint::unreachable_unchecked(),
        }
    }

    #[inline]
    pub fn i64(self) -> Option<i64> {
        match self {
            Self::I64(v) => Some(v),
            _ => None,
        }
    }

    /// # Safety
    /// - `self` must be `Self::I64`
    #[inline]
    pub unsafe fn i64_unchecked(self) -> i64 {
        debug_assert!(matches!(self, Self::I64(_)));
        match self {
            Value::I64(v) => v,
            _ => core::hint::unreachable_unchecked(),
        }
    }

    #[inline]
    pub fn f64(self) -> Option<f64> {
        match self {
            Self::F64(v) => Some(v),
            _ => None,
        }
    }

    /// # Safety
    /// - `self` must be `Self::F64`
    #[inline]
    pub unsafe fn f64_unchecked(self) -> f64 {
        debug_assert!(matches!(self, Self::F64(_)));
        match self {
            Value::F64(v) => v,
            _ => core::hint::unreachable_unchecked(),
        }
    }

    #[inline]
    pub fn bool(self) -> Option<bool> {
        match self {
            Self::Bool(v) => Some(v),
            _ => None,
        }
    }

    /// # Safety
    /// - `self` must be `Self::Bool`
    #[inline]
    pub unsafe fn bool_unchecked(self) -> bool {
        debug_assert!(matches!(self, Self::Bool(_)));
        match self {
            Value::Bool(v) => v,
            _ => core::hint::unreachable_unchecked(),
        }
    }
}

impl From<()> for Value {
    #[inline]
    fn from(_: ()) -> Self {
        Self::Unit
    }
}

impl From<i64> for Value {
    #[inline]
    fn from(value: i64) -> Self {
        Self::I64(value)
    }
}

impl From<f64> for Value {
    #[inline]
    fn from(value: f64) -> Self {
        Self::F64(value)
    }
}

impl From<bool> for Value {
    #[inline]
    fn from(value: bool) -> Self {
        Self::Bool(value)
    }
}

#[derive(Debug, Clone, Copy)]
#[repr(u64)]
pub enum Literal {
    JumpOffset(usize) = 0,
    I64(i64) = 1,
    F64(f64) = 2,
}

impl Literal {
    /// # Safety
    /// - `literal` must not be `Literal::Jmp`
    #[inline]
    pub unsafe fn into_value(self) -> Value {
        debug_assert!(!matches!(self, Literal::JumpOffset(_)));
        match self {
            Literal::JumpOffset(_) => unsafe { core::hint::unreachable_unchecked() },
            Literal::I64(v) => Value::I64(v),
            Literal::F64(v) => Value::F64(v),
        }
    }

    #[inline]
    pub fn tag(self) -> u64 {
        // SAFETY: Because `Self` is marked `repr(u64)`, its layout is a `repr(C)`
        // `union` between `repr(C)` structs, each of which has the `u64`
        // discriminant as its first field, so we can read the discriminant
        // without offsetting the pointer.
        unsafe {
            let ptr = &self as *const Self as *const u64;
            ptr.read()
        }
    }

    #[inline]
    pub fn i64(self) -> Option<i64> {
        match self {
            Self::I64(v) => Some(v),
            _ => None,
        }
    }

    /// # Safety
    /// - `self` must be `Literal::I64`
    #[inline]
    pub unsafe fn i64_unchecked(self) -> i64 {
        debug_assert!(!matches!(self, Literal::JumpOffset(_)));
        match self {
            Literal::I64(v) => v,
            _ => core::hint::unreachable_unchecked(),
        }
    }

    #[inline]
    pub fn f64(self) -> Option<f64> {
        match self {
            Self::F64(v) => Some(v),
            _ => None,
        }
    }

    /// # Safety
    /// - `self` must be `Literal::F64`
    #[inline]
    pub unsafe fn f64_unchecked(self) -> f64 {
        debug_assert!(!matches!(self, Literal::JumpOffset(_)));
        match self {
            Literal::F64(v) => v,
            _ => core::hint::unreachable_unchecked(),
        }
    }

    #[inline]
    pub fn jump_offset(self) -> Option<usize> {
        match self {
            Self::JumpOffset(v) => Some(v),
            _ => None,
        }
    }

    /// # Safety
    /// - `self` must be `Literal::JumpOffset`
    #[inline]
    pub unsafe fn jump_offset_unchecked(self) -> usize {
        debug_assert!(matches!(self, Literal::JumpOffset(_)));
        match self {
            Literal::JumpOffset(v) => v,
            _ => core::hint::unreachable_unchecked(),
        }
    }
}

impl Literal {
    #[inline]
    pub fn jmp(offset: usize) -> Self {
        Self::JumpOffset(offset)
    }
}

impl From<i64> for Literal {
    #[inline]
    fn from(value: i64) -> Self {
        Self::I64(value)
    }
}

impl From<f64> for Literal {
    #[inline]
    fn from(value: f64) -> Self {
        Self::F64(value)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn literal_value_tags_match() {
        assert_eq!(Literal::I64(0).tag(), Value::I64(0).tag());
        assert_eq!(Literal::F64(0.0).tag(), Value::F64(0.0).tag());
    }
}
