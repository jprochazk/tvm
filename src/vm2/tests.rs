use std::sync::Arc;

use super::value::Value;
use super::{asm, dispatch, Context, Error, Function};

fn context(functions: Vec<Arc<Function>>) -> Context {
    // always test with starting stack size of `1` to
    // properly exercise the stack growth logic
    Context::builder()
        .functions(functions)
        .initial_stack_size(1)
        .build()
}

#[test]
fn add_i64() {
    let functions = vec![Arc::new(Function::new(
        vec![
            asm::load_i16(0, 10),
            asm::load_i16(1, 10),
            asm::add_i64(0, 1, 0),
            asm::ret(),
        ],
        vec![],
        2,
    ))];

    let mut context = context(functions);
    dispatch(&mut context, 0).unwrap();

    assert!(
        matches!(context.ret(), Value::I64(20)),
        "{:?}",
        context.ret()
    );
    assert_eq!(context.stack_size(), 2);
}

#[test]
fn fibonacci() {
    let r0 = 0u8;
    let r1 = 1u8;
    let r2 = 2u8;
    let r3 = 3u8;
    let r4 = 4u8;
    let f1 = 1u16;

    // fib(3) + fib(7)

    let functions = vec![
        Arc::new(Function::new(
            vec![
                asm::load_i16(r1, 3i16),
                asm::call(r0, f1),
                asm::load_i16(r2, 7i16),
                asm::call(r1, f1),
                asm::add_i64(r0, r0, r1),
                asm::ret(),
            ],
            vec![],
            3,
        )),
        Arc::new(Function::new(
            vec![
                asm::load_i16(r2, 2i16),
                asm::cmp_lt_i64(r0, r1, r2),
                asm::jump_if_false(r0, 3),
                asm::mov(r0, r1),
                asm::jump(9),
                asm::load_i16(r4, 1i16),
                asm::sub_i64(r3, r1, r4),
                asm::call(r2, f1),
                asm::mov(r0, r2),
                asm::load_i16(r4, 2i16),
                asm::sub_i64(r3, r1, r4),
                asm::call(r2, f1),
                asm::add_i64(r0, r0, r2),
                asm::ret(),
            ],
            vec![],
            5,
        )),
    ];

    let mut context = context(functions);

    dispatch(&mut context, 0).unwrap();

    assert_eq!(
        context.ret().i64().unwrap(),
        iter_fib(3) + iter_fib(7),
        "{:?}",
        context.ret()
    );
}

fn iter_fib(n: i64) -> i64 {
    let mut a = 0;
    let mut b = 1;
    for _ in 0..n {
        let c = a + b;
        a = b;
        b = c;
    }

    a
}

#[test]
fn invalid_op() {
    let functions = vec![Arc::new(Function::new(vec![asm::invalid_op()], vec![], 0))];

    let mut context = context(functions);

    assert!(matches!(dispatch(&mut context, 0), Err(Error::InvalidOp)));
}
