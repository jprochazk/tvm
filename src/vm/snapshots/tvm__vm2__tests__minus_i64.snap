---
source: src/vm2/tests.rs
expression: "emit! (r#\"\n        let v = 1;\n        -v\n    \"#)"
---
I64(-1)
